File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const n = genres.length;
    
    // 1. 장르 별 재생횟수 구하기
    const genreMap = new Map();
    for (let i=0; i<n; i++) {
        genreMap.set(genres[i], (genreMap.get(genres[i])||0)+plays[i]);
    }
    
    const genreOrder = [];
    for (let [key, val] of genreMap) {
        genreOrder.push({genre: key, count: val});
    }
    genreOrder.sort((a,b) => b.count - a.count);
    
    // 2. 곡 별 재생횟수 구하기
    const playsOrder = [];
    for (let i=0; i<n; i++) {
        playsOrder.push({number: i, plays: plays[i]});
    }
    playsOrder.sort((a,b) => b.plays - a.plays);
    
    // 3. 전체 정렬
    const answer = [];
    for (let g of genreOrder) {
        let cnt = 0;
        for (let p of playsOrder) {
            if (genres[p.number]===g.genre) {
                answer.push(p.number);
                cnt++;
            }
            if (cnt >= 2) break;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    total = {}
    dic = {}
    for i,(g,p) in enumerate(zip(genres, plays)):
        if g in total:
            total[g]+=p
        else:
            total[g]=p
        if g in dic:
            dic[g].append((i,p))
        else:
            dic[g] = [(i,p)]
    
    answer = []
    for (k,v) in sorted(total.items(), key=lambda x:-x[1]):
        for (i,p) in sorted(dic[k], key=lambda x:(-x[1], x[0]))[:2]:
            answer.append(i)
    
    return answer